### ISSUE TITLE
Modify packet before it's sent

### ISSUE THREAD

**Issue by DaMatrix** [GeyserMC/MCProtocolLib#245]
I need to modify a packet before it's sent (namely, I need to change the UUID of a `LoginSuccessPacket`). Since I'm using PacketLib, I have no control over what UUID is used during the sending of said packet, and would like to change it after `session#send` has been called, but before it's actually sent. Is there a way I can do this, or no? Thanks!

**Comment by Steveice10** [comment]
There isn't. Why would you need to do this?

**Comment by Mareckoo01** [comment]
I would like that too, people usually don't want to put their login & password to some "random" application or don't even have spare account to do so.. it would be great if it would be possible to implement something like mcleaks api...

**Comment by Steveice10** [comment]
I don't quite understand. If you're concerned about logging in through MCProtocolLib you can just disable authentication, although you won't be able to connect to online mode servers obviously.

**Comment by phase** [comment]
@Mareckoo01 You can implement it yourself. Get the username & password from the api and connect to the server using MCPL.

**Comment by w-o-a-h** [comment]
^

**Comment by Mareckoo01** [comment]
I haven't looked into how auth & login works but alot of people use and trust this library so it should be ok..

@Steveice10 im not, but if people ask me if i could share my bot with them and i decide to share it, they might not want to put their login & pw in it to login..
or maybe they have only 1 minecraft account and then they can't play on the same server as the bot because the bot is on their acc..

@phase [this](https://mcleaks.net/apidoc) is what i would use, so i need to modify LoginStartPacket and EncryptionResponsePacket i think.

so instead of login & pw they can use [mcl alt token](https://mcleaks.net/get).

**Comment by DaMatrix** [comment]
I've fixed it by making my own class with the same canocial name as the class of the packet I was trying to modify, just tweaking a copypaste of the original code. But there should be a way to do this less hackily

**Comment by Steveice10** [comment]
I'm still not clear on what you're trying to do; if you're trying to login without a valid account, why not just disable authentication?

**Comment by Mareckoo01** [comment]
We are asking for a way to modify packets before they are sent(not by our code but by the library, i think @DaMatrix wants to modify _server to client_ packet, i want to modify _client to server_ packets) when i use
client.getSession().connect();
like this
```Java
@Override
public void packetSent(PacketSentEvent arg0) {
	if(arg0.getPacket() instanceof LoginStartPacket){ //Sent when client.getSession().connect(); (can't be modified)
		LoginStartPacket lsp = arg0.getPacket();
		lsp.setUsername("Nick"); // I want to change username
	}
	//Packet would be sent modified..
}
```

but i need to change not just packets but authentication too..

from mojang POST request authentication with username & password
to
mcl redeem - (user gets alt token from website) POST request to mcl with alt token

_**and**_

POST request to https://sessionserver.mojang.com/session/minecraft/join
to
POST request to http://auth.mcleaks.net/v1/joinserver

Quick links:
wiki.vg-
    [Protocol FAQ](http://wiki.vg/Protocol_FAQ)
    [Encryption Request](http://wiki.vg/Protocol#Encryption_Request)
    [Authentication](http://wiki.vg/Protocol_Encryption#Authentication)
mcl-
    [api](https://mcleaks.net/apidoc)
    [their forge mod(looking at source)](https://mcleaks.net/altloader) (skip to step3)



### **TL;DR**
im asking you to implement a way to edit packets before they are sent when i use client.getSession().connect();
im **not** asking you to implement mcl api or anything other
why? scroll up..

**Comment by Steveice10** [comment]
Yes, but as I keep saying, why do you need to modify packets to accomplish this when you can just disable authentication and give it whatever username you want to begin with?

**Comment by 0-x-2-2** [comment]
@Steveice10 change skins and we had a issue in this case with the uuid being wrong causing random client crashes and for things not working correctly

**Comment by 0-x-2-2** [comment]
so basically we want to modify packets before they are sent or before mcprotocollib does anything with received packets so atm we are doing really hackery things with asm and reflection.

**Comment by Steveice10** [comment]
...but once again, wouldn't changing the UUID in the login packets invalidate authentication?

**Comment by 0-x-2-2** [comment]
Maybe on the client to server but on the server to client seems to work fine.

**Comment by DaMatrix** [comment]
nope, it doesn't invalidate it. check out https://github.com/DaMatrix/Pork2b2tBot , that's where i'm using it at. i want the client to show up as having a specific UUID and name, BUT for security purposes, we are still using mojang auth.

**Comment by Steveice10** [comment]
https://github.com/Steveice10/PacketLib/commit/e8a86dafcb99f4952c423ee9a2bf66de95c7c245

Let me know if you encounter any further issues.

**Comment by 0-x-2-2** [comment]
Thanks!

**Comment by DaMatrix** [comment]
looks great, thanks!

**Comment by Mareckoo01** [comment]
Thanks!
I had to add Mareckoo01/PacketLib@98eabb9 and now it works great!

**Comment by Steveice10** [comment]
Thanks for pointing that out, should be fixed now.

**Comment by Mareckoo01** [comment]
If anyone would be looking for mcl login implementation, [here it is](https://gist.github.com/Mareckoo01/d67d1ab747a367f2c6143bb7c5c3ce5b)..
I didn't even need to edit packets.. well i'm not that good at coding because i do it just for fun..
